locator/token_metadata.cc

```diff
@@ -903,11 +912,15 @@ future<> token_metadata_impl::update_pending_ranges(
 }
 
 size_t token_metadata_impl::count_normal_token_owners() const {
+    return _nr_normal_token_owners;
+}
+
+void token_metadata_impl::update_normal_token_owners() {
     std::set<inet_address> eps;
     for (auto [t, ep]: _token_to_endpoint_map) {
         eps.insert(ep);
     }
-    return eps.size();
+    _nr_normal_token_owners = eps.size();
```

Comment:
I wonder if we can do one better and wrap `_token_to_endpoint_map` with a class, which intercepts all mutating methods and updates the number of token owners. This way we can avoid a developer modifying this structure, forgetting to call  the update method.

Reply:
token_metadata_impl _is_ such a class - it's a multi-index container of this stuff.

Reply:
It does seem that a more generic container struggles to come out here.
Something like `counting_values_unordered_map<key_type, value_type>` which offers `std::unordered_map` plus efficient counting of the number of different values present.

Reply:
We have multiple ways to mutate _token_to_endpoint_map in the code base. It is not trivial to wrap every mutate.
As Kamil mentioned, perhaps we can find a better container.

However, please keep in mind, I want to backport this to 4.6 for two users waiting for this in production. So I want to keep  the change minimum.